plugin/pulpcore/plugin/stages/profiler.py

```diff
@@ -0,0 +1,169 @@
+
+    The following statistics are computed for each Queue and the stage that it feeds into:
+
+        * waiting time - The number of seconds an item waited in the Queue for this stage.
+        * service time - The number of seconds an item received service in this stage.
+        * queue_length - The number of waiting items in the queue, measured before each new arrival.
+        * interarrival_time - The number of seconds since the previous arrival to this Queue.
+
+    See the :meth:`create_profile_db_and_connection()` docs for more info on the database tables and
+    layout.
+
+    Args:
+         stage_uuid (uuid.UUID): The uuid of the stage this ProfilingQueue delivers work into.
+         args (tuple): unused positional arguments
+         kwargs (dict): unused keyword arguments
+    """
+
+    def __init__(self, stage_uuid, *args, **kwargs):
+        self.last_arrival_time = time.time()
+        self.stage_uuid = stage_uuid
+        return super().__init__(*args, **kwargs)
+
+    def get_nowait(self):
+        """
+        Thinly wrap `asyncio.get_nowait` and record when get_nowait() operations happen.
+        """
+        item = super().get_nowait()
+        if item:
+            now = time.time()
+            if not hasattr(item, 'extra_data'):
+                # track stages that use QuerySet items too
+                item.extra_data = {}
+            item.extra_data['last_waiting_time'] = now - item.extra_data['last_put_time']
+            item.extra_data['last_get_time'] = now
+        return item
+
+    def put_nowait(self, item):
+        """
+        Thinly wrap `asyncio.put_nowait` happens and write statistics about them to the sqlite3 DB.
+
+        This method computes and writes the following statistics: waiting time, service time, queue
+        length, and interarrival time.
+        """
+        if item:
+            now = time.time()
+            if not hasattr(item, 'extra_data'):
+                # track stages that use QuerySet items too
+                item.extra_data = {}
+            try:
+                last_waiting_time = item.extra_data['last_waiting_time']
+            except KeyError:
+                pass
+            else:
+                service_time = now - item.extra_data['last_get_time']
+                sql = "INSERT INTO traffic (uuid, waiting_time, service_time) VALUES (" \
+                      "'{uuid}','{waiting_time}','{service_time}')"
+                formatted_sql = sql.format(
+                    uuid=self.stage_uuid, waiting_time=last_waiting_time, service_time=service_time
+                )
+                CONN.cursor().execute(formatted_sql)
+
+            interarrival_time = now - self.last_arrival_time
+            sql = "INSERT INTO system (uuid, length, interarrival_time) VALUES (" \
+                  "'{uuid}','{length}','{interarrival}')"
+            formatted_sql = sql.format(
+                uuid=self.stage_uuid, length=super().qsize(), interarrival=interarrival_time
```

Comment:
So the size is of the queue before the current item is put into it?

Reply:
Yes. I've tried to doc it in the docstring as such also. I think it's fine to measure before or after, it's the same, as long as it's clear so the data can be interpreted correctly.